proposals/0012-introduce-reactNativeMetadata.md

```diff
@@ -0,0 +1,467 @@
+
+## Future proof and extensibility
+
+We're versioning the schema using the `schema` key in the `reactNativeManifest` section. This allows us to evolve the schema in the future without breaking changes. We believe that the `schema` key should be a semver version string.
+
+We believe that the `reactNativeManifest` section should be extensible as we believe new capabilities and features can be added there.
+Specifically, we envision to evolve `reactNativeManifest` or `reactNativeManifest.capabilites` as follows:
+
+1. Each addition of a new capability [requires a new RFC](https://github.com/react-native-community/discussions-and-proposals#proposals) to be approved.
+2. Each RFC should define a new section in the `reactNativeManifest` with what's the intended use case.
+3. Each RFC should define what are the sensible defaults and should refrain from breaking changes.
+4. Each RFC should define the deprecation strategy for previous flags/features related to the same capability, if needed.
+
+## Complete Example
+
+Here we present a full example of how a `reactNativeManifest` section could look like for a library and an app.
+
+For an app the section will look as follows:
+
+```json
+{
+  "name": "my-awesome-app",
+  "version": "1.2.3",
+  "reactNativeManifest": {
+    "schema": "1.0.0",
+    "type": "app",
+    "capabilities": {
+      "hermes": {
+        "enabled": true
+      },
+      "newArchitecture": {
+        "enabled": true
+      },
+      "codegenConfig": {
+        "name": "AppSpecs",
+        "type": "all",
+        "jsSrcsDir": ".",
+        "android": {
+          "javaPackageName": "com.facebook.fbreact.specs"
+        }
+      }
+    }
+  }
+}
+```
+
+For a library the section will look as follows:
+
+```json
+{
+  "name": "my-awesome-library",
+  "version": "1.2.3",
+  "reactNativeManifest": {
+    "schema": "1.0.0",
+    "type": "library",
+    "capabilities": {
+      "newArchitecture": {
+        "enabled": true
+      },
+      "codegenConfig": {
+        "name": "LibrarySpecs",
+        "type": "all",
+        "jsSrcsDir": ".",
+        "android": {
+          "javaPackageName": "com.facebook.fbreact.specs"
+        }
+      }
+    }
+  }
+}
+```
+
+### JSON Schema
+
+The JSON Schema will be published as soon as we find an agreement on the key names and semantics.
+
+## Alternatives
+
+`reactNativeManifest` is, by design, fully original and new: there are other files that are used to define configurations for various aspect of react-native based projects (such as `react-native.config.js`, `app.json`, Expo's `app.config.js`, `expo-module.config.json`) but this one does **not** overlap with any of them. Explicitly, this one has the purpose of being filled only with capabilities and metadata to be read by package managers and build tools.
```

Comment:
In React Native template we have [`app.json`](https://github.com/facebook/react-native/blob/main/packages/react-native/template/app.json) file, and what's the role of this file in React Native CLI projects? It was added [here](https://github.com/facebook/react-native/pull/12162), for `npx react-native eject` purpose. `eject` command is a historical thing. What values should we put inside `app.json` in CLI projects, and what in `react-native.config.js` file.

Reply:
I am also curious about app.json and why it should continue to exist. It seems like a lot of the configuration there, such as the `expo` stuff, could be in the package.json.

@notbrent gave some more context here: https://github.com/react-native-community/cli/issues/1113#issuecomment-614794870

I realize that's a side-issue from this PR, though.

Reply:
Yeah, this issue focus on the parameters that React Native core needs to actually work and to be configured.

Personally, I believe that the `app.json` file, together with the `eject` commands should be removed completely as they are not really a thing as of today.

>  and what in `react-native.config.js` file.

The `react-native.config.js` is an additional and optional file that can be added to the project. It is independent from this proposal. As far as I understand, it can be used to add some configurations that are helpful to the CLI.
For that reason, if you need the info from the `app.json`, it should be merged into the `react-native.config.js` so that we reduce the number of files required by the template.

Reply:
Yeah, so `app.json` right now contains `name` which is then used inside `index.js` file to register root component, where should we put this value, after `app.json` deletion? I don't think that we should put this inside `react-native.config.js`, because then we need to add `react-native.config.js` to the template.

Reply:
Can't we just hardcode it into the `index.js` and update the CLI to replace that string? Unless it is used for something else, I don't understand what's the pro of externalize it.